Subject: [PATCH] git-instaweb: automatically select a port

If an explicit port has not been specified and the server supports
an idle timeout value and it fails to start because the automatically
selected port number is already in use, try several additional port
numbers until finding a free one.

Additionally, when using an automatically selected port number
always run the server with an idle timeout value if supported.

Signed-off-by: Kyle J. McKay <mackyle@gmail.com>

---
 git-instaweb.sh | 39 ++++++++++++++++++++++++++++++++++++---
 1 file changed, 36 insertions(+), 3 deletions(-)

diff --git a/git-instaweb.sh b/git-instaweb.sh
index b4bbf3b4..1da9fe8f 100755
--- a/git-instaweb.sh
+++ b/git-instaweb.sh
@@ -30,8 +30,11 @@ httpd="$(git config --get instaweb.httpd)"
 root="$(git config --get instaweb.gitwebdir)"
 port=$(git config --get instaweb.port)
 module_path="$(git config --get instaweb.modulepath)"
+timeout=$(git config --int --get instaweb.timeout 2>/dev/null || :)
+case "$timeout" in -*) timeout=;; esac
 action="browse"
 hostname="127.0.0.1"
+isdefport=
 
 conf="$GIT_DIR/gitweb/httpd.conf"
 
@@ -44,7 +47,10 @@ test -z "$httpd" && httpd='lighttpd -f'
 test -z "$root" && root='@@GITWEBDIR@@'
 
 # any untaken local port will do...
-test -z "$port" && port=1234
+test -z "$port" && isdefport=1 && port=1234
+
+# default timeout is 9h6m7s
+test -z "$timeout" && timeout=32767
 
 select_hostname () {
 	case "$httpd" in
@@ -105,6 +111,15 @@ resolve_full_httpd () {
 	fi
 }
 
+is_port_in_use () {
+	case "$1" in *[Bb]"ind to port"*)
+		case "$1" in *[Aa]"ddress already in use"*)
+			return 0
+		esac
+	esac
+	return 1
+}
+
 start_httpd () {
 	if test -s "$fqgitdir/pid"; then
 		say "Instance already running. Restarting..."
@@ -121,6 +136,15 @@ start_httpd () {
 	test -f "$fqgitdir/gitweb/git-browser.conf" || gitbrowser_conf
 	test -f "$fqgitdir/gitweb/gitweb_config.perl" || gitweb_conf
 
+	# check for timeout support if lighttpd and using default port
+	timeoutopt=
+	if test -n "$isdefport"; then case "$httpd" in *lighttpd*)
+		errstr="$(${full_httpd%%-f*} -i -1 2>&1 || :)"
+		case "$errstr" in *[Ii]"nvalid idle timeout value"*)
+			timeoutopt="-i $timeout"
+		esac
+	esac; fi
+
 	# don't quote $full_httpd, there can be arguments to it (-f)
 	case "$httpd" in
 	*mongoose*|*plackup*)
@@ -139,8 +163,17 @@ $pid
 EOF
 		;;
 	*)
-		$full_httpd "$conf"
-		if test $? != 0; then
+		errstr="$($full_httpd "$conf" $timeoutopt 2>&1)"; err=$?
+		if test $err != 0 && test -n "$timeoutopt" && is_port_in_use "$errstr"; then
+			attempts=25
+			while attempts=$(( $attempts - 1 )) && port=$(( $port + 1)) && test $attempts != 0; do
+				configure_httpd
+				errstr="$($full_httpd "$conf" $timeoutopt 2>&1)"; err=$?
+				test $err != 0 && is_port_in_use "$errstr" || break
+			done
+		fi
+		test -z "$errstr" || printf '%s\n' "$errstr"
+		if test $err != 0; then
 			echo "Could not execute http daemon $httpd."
 			exit 1
 		fi
---
